src/fluree/server/components/subscriptions.clj

```diff
@@ -0,0 +1,219 @@
+  (let [{:keys [chan ws]} (get-in @sub-atom [:subs sub-id])]
+    (close-chan chan nil)
+    (try (ws/close! ws)
+         (catch Exception _ :ignore))
+    (swap! subscriptions update :subs dissoc sub-id)))
+
+(defn subscribe-ledger
+  "Subscribe to ledger"
+  [{:keys [sub-atom] :as subscriptions} sub-id {:keys [ledger serialization] :as request}]
+  (log/debug "Subscribe websocket request by sub-id" sub-id "request:" request)
+  (swap! sub-atom update-in [:subs sub-id]
+         (fn [existing-sub]
+           (if existing-sub
+             (assoc-in existing-sub [:ledgers ledger] {:serialization serialization})
+             (log/info "Existing subscription for sub-id: " sub-id "cannot be found, assume just closed.")))))
+
+(defn unsubscribe-ledger
+  "Unsubscribe from ledger"
+  [{:keys [sub-atom] :as subscriptions} sub-id {:keys [ledger] :as request}]
+  (log/debug "Unsubscribe websocket request by sub-id" sub-id "request:" request)
+  (swap! sub-atom update-in [:subs sub-id :ledgers]
+         (fn [ledger-subs]
+           (if ledger-subs
+             (dissoc ledger-subs ledger)
+             (log/info "Existing subscription for sub-id: " sub-id "cannot be found, assume just closed.")))))
+
+(defn send-message
+  "Sends a message to an individual socket. If an exception occurs,
+  closes the socket and removes it from the subscription map."
+  [{:keys [sub-atom] :as subscriptions} sub-id message]
+  (let [ws (get-in @sub-atom [:subs sub-id :ws])]
+    (try
+      (ws/send! ws message)
+      (catch IOException _
+        (log/info "Websocket channel closed (java.io.IOException) for sub-id: " sub-id))
+      (catch ClosedChannelException _
+        (log/info "Websocket channel closed for sub-id: " sub-id)
+        (close-subscription subscriptions sub-id nil nil))
+      (catch Exception e
+        (log/error e "Error sending message to websocket subscriber: " sub-id)
+        (close-subscription subscriptions sub-id nil nil)))))
+
+
+(defn send-ledger-message
+  "Sends a message only to subscriptions that subscribed to a ledger.
+  Note the message format should be a JSON-stringified map in the form of:
+  {action: commit, ledger: myledger, data: {...}}"
+  [{:keys [sub-atom] :as subscriptions} ledger-id message]
+  (log/debug "Sending ledger message for:" ledger-id "with message:" message)
+  (let [{:keys [closed? subs]} @sub-atom]
+    (when-not closed?
+      (doseq [[sub-id {:keys [ws ledgers]}] subs]
+        (when (contains? ledgers ledger-id)
+          ;; TODO - use ledger subscription opts to filter out messages, permission
+          (send-message subscriptions sub-id message))))))
+
+
+(defn send-message-to-all
+  "Sends a message to all subscriptions. Message sent as JSON stringified map
+  in the form: {action: commit, ledger: my/ledger, data: {...}}"
+  [{:keys [sub-atom] :as subscriptions} action ledger-alias data]
+  (log/warn "SEND MESSAGE TO ALL subscriptions: " subscriptions)
```

Comment:
Why is this logged at `warn` level?

Reply:
This was a debugging message, I'll remove.